### PR TITLE
Data Science Persona

### PR DESCRIPTION
**Pull request by jonahjung22**
## Problem

Users need a persona that can make data science-related intelligent decisions about analysis approaches based on context, understand and analyze Jupyter notebooks, provide targeted actionable insights rather than generic responses, and handle complex data science workflows with proper error handling and fallbacks.

## Solution

Built an intelligent PocketFlow data science persona that uses LLM reasoning for decision-making and workflow orchestration. The solution implements a three-node architecture with DecideAction, DataAnalysis, and CompleteAnalysis nodes featuring robust YAML parsing, context integration, and adaptive routing to deliver context-aware data science analysis that intelligently chooses between focused analysis and comprehensive reviews based on user intent and available context.

## Changes

- **Code**: Implemented PocketFlow-based agent architecture with DecideAction, DataAnalysis, and CompleteAnalysis nodes, added robust YAML decision parsing with multiple fallback strategies for LLM response handling, created intelligent notebook detection and context loading system, built comprehensive error handling and logging throughout the workflow, and integrated AWS Bedrock for LLM decision-making with graceful fallbacks. 

- **Tests**: Added tests for YAML parsing edge cases and malformed LLM responses, created integration tests for notebook detection and context loading, implemented workflow routing tests for all decision action types, and added error handling tests for missing files and configuration issues.

- **Docs**: Created comprehensive README with architecture overview, usage examples, and troubleshooting guide, added API documentation for all node classes and methods, included configuration guide for AWS Bedrock setup, and documented performance characteristics and system requirements.

## Testing Instructions

Set up AWS Bedrock credentials in Jupyter AI configuration, create a test notebook with data science content, create a repo_context.md file in your working directory, test basic usage with "@DataSciencePersona analyze my data", test specific notebook analysis with "@DataSciencePersona notebook: path/to/file.ipynb help me improve my model", verify error handling by testing with missing files and malformed configurations, and check logging output for decision reasoning and workflow routing to ensure the three-step architecture is working correctly.

### CODE REVIEW

**Review by Zsailer — COMMENTED** [review]
## Overall Review

This PR introduces a sophisticated data science persona with PocketFlow agent architecture. The implementation shows good architectural thinking with decision-making nodes and comprehensive error handling. However, there are several areas that need improvement before merging.

**Review by Zsailer — COMMENTED** [review]
# Code Quality Issues to Address:

## 1. **Error Handling & Logging**
- Lines 59-73 in `agent.py`: Complex nested response parsing logic should be extracted into a helper method
- Inconsistent error handling patterns - some methods use try/catch, others don't
- Logger setup needs improvement with proper formatting and levels

## 2. **Code Structure & Maintainability**
- The `DecideAction` class is quite large (200+ lines) and has multiple responsibilities
- YAML parsing logic (lines 119-186) should be extracted into a separate utility class
- Too many magic strings for action types - consider using an enum or constants

## 3. **Configuration & Dependencies**
- Hard-coded dependency on AWS Bedrock without proper fallback documentation
- Missing validation for required configuration parameters
- No clear dependency injection pattern for the model client

## 4. **Performance & Resource Management**
- No caching for repeated context loading
- Large notebook content might cause memory issues (only 2MB limit mentioned)
- No timeout handling for LLM calls

## 5. **Testing & Documentation**
- README is comprehensive but lacks troubleshooting for common failure scenarios
- No unit tests visible in the PR
- Missing docstrings for many private methods
EOF < /dev/null

File: jupyter_ai_personas/data_science_persona/agent.py [hunk 1]
```diff
@@ -0,0 +1,1367 @@
+"""
+PocketFlow Agent for Data Science Analysis
+
+A sophisticated agent that can reason about data science tasks and decide
+on appropriate actions based on user queries, repo context, and notebook content.
+"""
+
+import logging
+import yaml
+from pathlib import Path
```
Comment (Zsailer) [inline jupyter_ai_personas/data_science_persona/agent.py:10]: **Extract Complex Response Parsing**

This nested response parsing logic is complex and hard to maintain. Consider extracting it to a dedicated method:

```python
def _extract_response_content(self, response):
    """Extract content from various response formats"""
    if hasattr(response, 'content'):
        return response.content
    elif isinstance(response, dict):
        return self._extract_from_dict_response(response)
    else:
        return str(response)
```

This would make the `exec` method cleaner and easier to test.

File: jupyter_ai_personas/data_science_persona/agent.py [hunk 2]
```diff
@@ -0,0 +1,1367 @@
...
+        super().__init__()
+        self.model_client = model_client
+    
+    def prep(self, shared):
+        """Prepare context for decision making"""
+        return {
+            "user_query": shared.get("user_query", ""),
+            "repo_context": shared.get("repo_context", ""),
+            "notebook_content": shared.get("notebook_content", ""),
+            "notebook_path": shared.get("notebook_path", ""),
```
Comment (Zsailer) [inline jupyter_ai_personas/data_science_persona/agent.py]: **Consider Using Enum for Action Types**

Instead of hardcoded action strings, consider using an enum for better type safety:

```python
from enum import Enum

class ActionType(str, Enum):
    ANALYZE_DATA = "analyze_data"
    GENERATE_CODE = "generate_code"
    EXPLAIN_CONCEPT = "explain_concept"
    FIND_ISSUES = "find_issues"
    CREATE_VISUALIZATION = "create_visualization"
    OPTIMIZE_MODEL = "optimize_model"
    DEBUG_CODE = "debug_code"
    COMPLETE_ANALYSIS = "complete_analysis"
    GREETING = "greeting"
```

This prevents typos and makes the code more maintainable.

File: jupyter_ai_personas/data_science_persona/agent.py (hunk 1 above)
Comment (Zsailer) [inline jupyter_ai_personas/data_science_persona/agent.py:2]: **Extract Complex Response Parsing**

This nested response parsing logic is complex and hard to maintain. Consider extracting it to a dedicated method:

```python
def _extract_response_content(self, response):
    """Extract content from various response formats"""
    if hasattr(response, 'content'):
        return response.content
    elif isinstance(response, dict):
        return self._extract_from_dict_response(response)
    else:
        return str(response)
```

This would make the `exec` method cleaner and easier to test.

**Comment by Zsailer** [comment]
Playing around with AI for reviewing PRs :). Feel free to ignore the comments above for now.

File: jupyter_ai_personas/data_science_persona/README.md [hunk 3]
```diff
@@ -0,0 +1,412 @@
...
+print("🏆 AutoGluon Model Leaderboard:")
+print(leaderboard.head(10))
+
+best_model = leaderboard.iloc[0]['model']
+best_score = leaderboard.iloc[0]['score_val']
+print(f"\\n🥇 BEST MODEL: {best_model}")
+print(f"📊 BEST SCORE: {best_score:.4f}")
+```
+"""
+```
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/README.md]: Remove these two lines 107-108 as they mess up the formatting of the page.

File: jupyter_ai_personas/data_science_persona/README.md [hunk 4]
```diff
@@ -0,0 +1,412 @@
...
+    prediction_length=7,
+    path='./autogluon_models/timeseries_model'
+).fit(
+    TimeSeriesDataFrame(ts_data_formatted),
+    time_limit=120,
+    presets='best_quality'
+)
+```
+"""
+```
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/README.md]: Remove these lines 159-160 as they spoil the markdown formatting.

File: jupyter_ai_personas/data_science_persona/README.md [hunk 5]
```diff
@@ -0,0 +1,412 @@
...
+    product_data,
+    time_limit=240,  # Multimodal typically needs more time
+    presets='best_quality'
+)
+
+print("✅ Multimodal training completed!")
+print("🎯 Model handles text, images, and numerical data automatically!")
+```
+"""
+```
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/README.md]: Remove these two lines

File: jupyter_ai_personas/data_science_persona/agent.py [hunk 6]
```diff
@@ -0,0 +1,497 @@
+import logging
+import re
+from pathlib import Path
+from typing import Dict, Any
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/agent.py]: Dict, Any are not accessed anywhere, remove?

File: jupyter_ai_personas/data_science_persona/agent.py [hunk 7]
```diff
@@ -0,0 +1,497 @@
...
+    )
+except ImportError as e:
+    logging.error(f"Failed to import required modules: {e}")
+    raise ImportError(f"Missing dependencies for DataScienceAgent: {e}") from e
+
+logger = logging.getLogger(__name__)
+
+class DataScienceAgent(Flow):
+    """
+    PocketFlow Agent that coordiantes the workflow for Data Science Analysis
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/agent.py]: typo of  "coordinates"

File: jupyter_ai_personas/data_science_persona/agent.py [hunk 8]
```diff
@@ -0,0 +1,497 @@
...
+            
+            # Determine primary domain
+            primary_domain = max(domain_scores.items(), key=lambda x: x[1])[0]
+            suggested_domains = [domain for domain, score in domain_scores.items() if score > 0]
+            
+            analysis_result.update({
+                "primary_domain": primary_domain,
+                "suggested_domains": suggested_domains,
+                "domain_scores": domain_scores
+            })
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/agent.py]: Is there a way to replace this logic using `df.dtypes` for each column? You may be able to do this in very few lines of code once you know what the `dtype` is for each column.

File: jupyter_ai_personas/data_science_persona/agent.py [hunk 9]
```diff
@@ -0,0 +1,497 @@
...
+                "success": False,
+                "error": f"Analysis failed: {e}",
+                "primary_domain": "tabular",
+                "suggested_domains": ["tabular"]
+            }
+    
+    def _load_repo_context(self):
+        """Load repository context from repo_context.md"""
+        try:
+            repo_path = Path.cwd() / "repo_context.md"
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/agent.py:289]: From here on, put comments at the head of each function so it is clear what each function does.

File: jupyter_ai_personas/data_science_persona/autogluon_tool.py [hunk 10]
```diff
@@ -0,0 +1,614 @@
...
+        if target_candidates:
+            return target_candidates[0]
+        
+        # For time series, often the last numeric column is the target
+        numeric_cols = df.select_dtypes(include=['number']).columns.tolist()
+        if numeric_cols:
+            return numeric_cols[-1]  # Use last numeric column
+        
+        # Fallback to last column
+        return df.columns[-1] if len(df.columns) > 0 else 'target'
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/autogluon_tool.py]: Not sure this is the best fallback. Maybe call a LLM and let it decide? Prompt it to return the column most similar to the list in `common_targets`?

File: jupyter_ai_personas/data_science_persona/autogluon_tool.py [hunk 11]
```diff
@@ -0,0 +1,614 @@
...
+        """Train actual AutoGluon time series model and return results."""
+        try:
+            if not self.availability.get("timeseries", False):
+                return {
+                    "success": False,
+                    "error": "AutoGluon timeseries not available. Install with: pip install autogluon.timeseries"
+                }
+            
+            from autogluon.timeseries import TimeSeriesDataFrame, TimeSeriesPredictor
+            import pandas as pd
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/autogluon_tool.py]: move import to the top of the module?

File: jupyter_ai_personas/data_science_persona/autogluon_tool.py [hunk 12]
```diff
@@ -0,0 +1,614 @@
...
+print(f"Generated {{len(forecasts)}} forecast steps")
+print(f"Target: {{actual_target}}")
+print(f"Prediction Length: {prediction_length} steps")
+
+print(f"\\nSample Forecasts:")
+print(forecasts.head(10))
+
+print(f"\\nModel Selection:")
+print("AutoGluon automatically selected the best performing model from the ensemble")
+print("The WeightedEnsemble combines multiple models for optimal performance")"""
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/autogluon_tool.py]: Indent the prompt cleanly

File: jupyter_ai_personas/data_science_persona/autogluon_tool.py [hunk 13]
```diff
@@ -0,0 +1,614 @@
...
+# 🥇 BEST MODEL INFORMATION
+best_model = leaderboard.iloc[0]['model']
+best_score = leaderboard.iloc[0]['score_val']
+print(f"\\n🥇 BEST MODEL: {{best_model}}")
+print(f"📊 BEST SCORE: {{best_score:.4f}}")
+
+# 📈 DETAILED RANKING
+print("\\n📈 Top 5 Models Ranking:")
+for i, row in leaderboard.head(5).iterrows():
+    print(f"{{i+1:2d}}. {{row['model']:25s}} | Score: {{row['score_val']:.4f}} | Time: {{row['fit_time']:.1f}}s")"""
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/autogluon_tool.py]: Indent the prompt to make it tidy

File: jupyter_ai_personas/data_science_persona/autogluon_tool.py [hunk 14]
```diff
@@ -0,0 +1,614 @@
...
+    label=actual_target,
+    path='./autogluon_models/multimodal_model'
+).fit(
+    train_data,
+    time_limit={self.default_time_limit * 2},  # Multimodal typically needs more time
+    presets='best_quality'
+)
+
+print(f"✅ Multimodal training completed for {{actual_target}}!")
+print("Model handles text, images, and numerical data automatically!")"""
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/autogluon_tool.py]: Indent prompt string

**Review by srdas — COMMENTED** [review]
[1] When attempting to install from scratch, either for `all` personas or just the  `data_science` persona, the following install error occurs:
image
Note that this error is not encountered with any of the other personas, so there is some issue with the install. As of now, I cannot test the persona. 

[2] The main `README.md` file for `jupyter-ai-personas` needs to be updated for the table to show this persona as well. https://github.com/jupyter-ai-contrib/jupyter-ai-personas/blob/main/README.md

File: jupyter_ai_personas/data_science_persona/file_reader_tool.py [hunk 15]
```diff
@@ -0,0 +1,170 @@
+import json
+import os
+from typing import Dict, Any, List, Optional
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/file_reader_tool.py]: `Optional` is not accessed (drop it)?

File: jupyter_ai_personas/data_science_persona/nodes.py [hunk 16]
```diff
@@ -0,0 +1,947 @@
+import logging
+import yaml
+import re
+from pathlib import Path
+from typing import Dict, Any
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/nodes.py]: These do not appear to be used anywhere?

File: jupyter_ai_personas/data_science_persona/nodes.py [hunk 17]
```diff
@@ -0,0 +1,947 @@
...
+    """
+    Decision-making node that analyzes the user query and context
+    to determine the appropriate action for data science analysis.
+    """
+    
+    def __init__(self, model_client=None):
+        super().__init__()
+        self.model_client = model_client
+    
+    def prep(self, shared):
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/nodes.py:24]: here and for all functions, add comments to explain what each module does?

File: jupyter_ai_personas/data_science_persona/nodes.py [hunk 18]
```diff
@@ -0,0 +1,947 @@
...
+## 🚀 **Getting Started:**
+Just tell me what you'd like to work on! For example:
+- "Analyze my sales data for trends"
+- "Help me train a classification model" 
+- "Debug my notebook: notebook_name.ipynb"
+- "Optimize my data preprocessing pipeline"
+
+I'll automatically read your repository context and notebook content to provide targeted, actionable recommendations.
+
+What would you like to explore today? 🎯"""
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/nodes.py:302]: Indent for neatness?

File: jupyter_ai_personas/data_science_persona/persona.py [hunk 19]
```diff
@@ -0,0 +1,223 @@
+import logging
+from typing import Dict, Any, AsyncGenerator, Optional
```
Comment (srdas) [inline jupyter_ai_personas/data_science_persona/persona.py]: `Optional` is not accessed.

File: jupyter_ai_personas/data_science_persona/agent.py (hunk 8 above)
Comment (jonahjung22) [reply jupyter_ai_personas/data_science_persona/agent.py]: I agree I think this method is inefficient and sometimes inaccurate, I am going to pass the defined data to the model and have it return the domain type.

File: jupyter_ai_personas/data_science_persona/autogluon_tool.py (hunk 10 above)
Comment (jonahjung22) [reply jupyter_ai_personas/data_science_persona/autogluon_tool.py]: I have been exploring this option of implementing an LLM but I think the process of calling the LLM may be too computationally expensive for this case. This tool is providing sample code for the user, so they should be able to change a variable to their ideal target values.

File: jupyter_ai_personas/data_science_persona/autogluon_tool.py (hunk 12 above)
Comment (jonahjung22) [reply jupyter_ai_personas/data_science_persona/autogluon_tool.py]: The indentation affects the formatting of the code snippet, as much as I would like to indent it as well.
